In [1]:
%conda install matplotlib -y

import pandas as pd
import numpy as np
import statsmodels.api as sm

from sklearn import datasets
from sklearn.linear_model import LinearRegression

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 23.3.1
  latest version: 23.10.0

Please update conda by running

    $ conda update -n base -c defaults conda

Or to minimize the number of packages updated during conda update use

     conda install conda=23.10.0



# All requested packages already installed.


Note: you may need to restart the kernel to use updated packages.


# Simple Linear Regression - practice

## Simple Linear Regression 

- In regression analysis, the variable to be explained is the dependent variable.
- In regression analysis, the variable used to describe is an independent variable.
- Simple linear regression uses the least squares method to find the most appropriate model.
- The higher the coefficient of determination, the better the model explains the phenomenon.

### Practice Simple Linear Regression
The following are some of the data regarding diabetic patients. Propose a model to explain diabetes levels using BMI, and test properly.

In [2]:
dia = datasets.load_diabetes()

dia = pd.DataFrame(dia['data'], index=dia['target'], columns=dia['feature_names'])
display(dia)

age       sex       bmi        bp        s1        s2        s3  \
151.0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
75.0  -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
141.0  0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
206.0 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
135.0  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   
...         ...       ...       ...       ...       ...       ...       ...   
178.0  0.041708  0.050680  0.019662  0.059744 -0.005697 -0.002566 -0.028674   
104.0 -0.005515  0.050680 -0.015906 -0.067642  0.049341  0.079165 -0.028674   
132.0  0.041708  0.050680 -0.015906  0.017293 -0.037344 -0.013840 -0.024993   
220.0 -0.045472 -0.044642  0.039062  0.001215  0.016318  0.015283 -0.028674   
57.0  -0.045472 -0.044642 -0.073030 -0.081413  0.083740  0.027809  0.173816   

             s4        s5        s6  
151.0 -0.002592  0.019907 -0.017646  
75.0  -0.039493 -0.068332 -0.092204  
141.0 -0.002592  0.002861 -0.025930  
206.0  0.034309  0.022688 -0.009362  
135.0 -0.002592 -0.031988 -0.046641  
...         ...       ...       ...  
178.0 -0.002592  0.031193  0.007207  
104.0  0.034309 -0.018114  0.044485  
132.0 -0.011080 -0.046883  0.015491  
220.0  0.026560  0.044529 -0.025930  
57.0  -0.039493 -0.004222  0.003064  

[442 rows x 10 columns]

After extracting only BMI and diabetes levels, the model calculated using the least squares method is as follows.

In [3]:
dia = dia['bmi']
display(dia)

lr = LinearRegression()

lr.fit(dia.values.reshape(-1, 1), dia.index.to_numpy().reshape(-1, 1))

print(f"Model: y = {lr.intercept_} + {lr.coef_[0]}X + e")

151.0    0.061696
75.0    -0.051474
141.0    0.044451
206.0   -0.011595
135.0   -0.036385
           ...   
178.0    0.019662
104.0   -0.015906
132.0   -0.015906
220.0    0.039062
57.0    -0.073030
Name: bmi, Length: 442, dtype: float64

Model: y = [152.13348416] + [949.43526038]X + e


Null Hypothesis: There is no linear relationship between independent and dependent variables.  
Alternative Hypothesis: There is a linear relationship between independent and dependent variables.

In [4]:
results = sm.OLS(dia.index.to_numpy(), sm.add_constant(dia.values)).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.344
Model:                            OLS   Adj. R-squared:                  0.342
Method:                 Least Squares   F-statistic:                     230.7
Date:                Sat, 18 Nov 2023   Prob (F-statistic):           3.47e-42
Time:                        18:35:04   Log-Likelihood:                -2454.0
No. Observations:                 442   AIC:                             4912.
Df Residuals:                     440   BIC:                             4920.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        152.1335      2.974     51.162      0.000     146.289     157.978
x1           949.4353     62.515     15.187      0.000     826.570    1072.301
==============================================================================
Omnibus:                       11.674   Durbin-Watson:                   1.848
Prob(Omnibus):                  0.003   Jarque-Bera (JB):                7.310
Skew:                           0.156   Prob(JB):                       0.0259
Kurtosis:                       2.453   Cond. No.                         21.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

interpretation  
First, the coefficient of determination is 0.344, so our model describes some of the data.  
In addition, since the p-value for x1 is 0.000, there is sufficient evidence to reject the null hypothesis at the 99.9% confidence level.  
=> There is a linear relationship between the two variables! However, need be aware of the low coefficient of determination.